var/spack/repos/builtin/packages/babelstream/package.py

```diff
@@ -38,67 +45,184 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
+    variant("std", default=False, description="Enable STD support")
 
     # Some models need to have the programming model abstraction downloaded -
     # this variant enables a path to be provided.
     variant("dir", values=str, default="none", description="Enable Directory support")
+    variant(
+        "sycl2020_submodel",
+        values=("usm", "acc"),
+        when="+sycl2020",
+        default="usm",
+        description="SYCL2020 -> choose between usm and acc methods",
+    )
+    variant(
+        "std_submodel",
+        values=("data", "indices", "ranges"),
+        when="+std",
+        default="data",
+        description="STD -> choose between data, indices and ranges models",
+    )
+    variant(
+        "sycl_opencl_path",
+        values=str,
+        when="+sycl",
+        default="none",
+        description="[ComputeCpp only] Path to OpenCL library, usually called libOpenCL.so",
+    )
 
-    # Kokkos conflict and variant
-    conflicts(
-        "dir=none", when="+kokkos", msg="KOKKKOS requires architecture to be specfied by dir="
+    sycl_compiler_implementations = [
+        "oneapi-icpx",
+        "oneapi-clang",
+        "dpcpp",
+        "hipsycl",
+        "computecpp",
+        "none",
+    ]  # list them here, with the default first
+    sycl_compiler_implementations_description = "Compile using the specified SYCL compiler implementation\
+           Supported values are \
+           ONEAPI-ICPX  - icpx as a standalone compiler \
+           ONEAPI-Clang - oneAPI's Clang driver (enabled via `source /opt/intel/oneapi/setvars.sh  --include-intel-llvm`)\
+           DPCPP        - dpc++ as a standalone compiler (https://github.com/intel/llvm)\
+           HIPSYCL      - hipSYCL compiler (https://github.com/illuhad/hipSYCL)s\
+           COMPUTECPP   - ComputeCpp compiler (https://developer.codeplay.com/products/computecpp/ce/home)"
+    variant(
+        "sycl_compiler_implementation",
+        values=sycl_compiler_implementations,
+        default=sycl_compiler_implementations[0],
+        when="+sycl",
+        description=sycl_compiler_implementations_description,
+    )
+    variant(
+        "sycl_compiler_implementation",
+        values=sycl_compiler_implementations,
+        default=sycl_compiler_implementations[0],
+        when="+sycl2020",
+        description=sycl_compiler_implementations_description,
+    )
+
+    variant(
+        "thrust_submodel",
+        values=("cuda", "rocm"),
+        default="cuda",
+        when="+thrust",
+        description="Which THRUST implementation to use, supported options include option= \
+            - CUDA (via https://github.com/NVIDIA/thrust)\
+            - ROCM (via https://github.com/ROCmSoftwarePlatform/rocThrust)",
     )
+    variant(
+        "thrust_backend",
+        values=("cuda", "omp", "tbb"),
+        default="cuda",
+        when="+thrust",
+        description="Which THRUST implementation to use, supported options include option",
+    )
```

Comment:
I *think* this should be a virtual as well, e.g. `thrust-api` so the user can pick their thrust implementation. This and other suggestions above would simplify all the conditional logic in your dependencies above.

Reply:
@tgamblin  it just sets `args.append("-DBACKEND=" + self.spec.variants["thrust_backend"].value.upper())` so I think it is simple that doesn't need any extra effort on spack package ( taken care by the CMake)